go/libraries/doltcore/sqle/mergeable_indexes_test.go

```diff
@@ -0,0 +1,938 @@
+// Copyright 2020 Dolthub, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package sqle
+
+import (
+	"context"
+	"fmt"
+	"testing"
+
+	"github.com/dolthub/go-mysql-server/sql"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+
+	"github.com/dolthub/dolt/go/libraries/doltcore/sqle/lookup"
+)
+
+// This tests mergeable indexes by using the SQL engine and intercepting specific calls. This way, we can verify that
+// the engine is intersecting and combining the proper number of lookups, and we can also examine the ranges before
+// they're converted into a format that Noms understands to verify that they were handled correctly. Lastly, we ensure
+// that the final output is as expected.
+func TestMergeableIndexes(t *testing.T) {
```

Comment:
These tests are great, but you're missing an important class of them. I want to see things like

v1 > 1 AND v1 < 5 OR v1 > 6 AND v1 < 9

And variations.

Also things like:

(v1 > 1 AND v1 < 5 OR v1 > 6 AND v1 < 9) AND v1 > 3

Things where you're combining complex sets of predicates together with and AND or OR.

Reply:
Conspicuously missing: testing use of index for IN clauses (v1 in (1,2,3)). If that doesn't work in the engine yet, then skip the test, make it work, and then unskip it

Reply:
So right now I have (replacing X for =, >, <, etc.):

X1
X1 OR X2
X1 AND X2
X1 OR X2 OR X3
X1 OR X2 AND X3
X1 AND X2 OR X3
X1 AND X2 AND X3

So you're saying I should add:

X1 OR X2 OR X3 OR X4
X1 OR X2 OR X3 AND X4
X1 OR X2 AND X3 OR X4
X1 OR X2 AND X3 AND X4
X1 AND X2 OR X3 OR X4
X1 AND X2 OR X3 AND X4
X1 AND X2 AND X3 OR X4
X1 AND X2 AND X3 AND X4
X1 OR X2 OR X3 OR X4 OR X5
X1 OR X2 OR X3 OR X4 AND X5
X1 OR X2 OR X3 AND X4 OR X5
X1 OR X2 OR X3 AND X4 AND X5
X1 OR X2 AND X3 OR X4 OR X5
X1 OR X2 AND X3 OR X4 AND X5
X1 OR X2 AND X3 AND X4 OR X5
X1 OR X2 AND X3 AND X4 AND X5
X1 AND X2 OR X3 OR X4 OR X5
X1 AND X2 OR X3 OR X4 AND X5
X1 AND X2 OR X3 AND X4 OR X5
X1 AND X2 OR X3 AND X4 AND X5
X1 AND X2 AND X3 OR X4 OR X5
X1 AND X2 AND X3 OR X4 AND X5
X1 AND X2 AND X3 AND X4 OR X5
X1 AND X2 AND X3 AND X4 AND X5

Or I should just add a few of them to verify that they're working as expected (not needing such an exhaustive amount)?

Also, I didn't even think about `IN` clauses, I'll get those added.

Reply:
What I was expecting to see (and didn't) was a where clause that resulted in two closed ranges on the lookup. That's the main thing I want to see covered.

Reply:
Ahh gotcha. I can add a few that results in those.